src/game/savegame_bson.c

```diff
@@ -969,14 +1043,25 @@ bool Savegame_BSON_LoadFromFile(MYFILE *fp, GAME_INFO *game_info)
         goto cleanup;
     }
 
-    if (!Savegame_BSON_LoadStartInfo(
-            json_object_get_array(root_obj, "start_info"), game_info)) {
+    if (!Savegame_BSON_LoadResumeInfo(
+            json_object_get_array(root_obj, "start_info"), game_info->start)) {
+        LOG_DEBUG("Failed to load start_info properly");
         goto cleanup;
     }
 
-    if (!Savegame_BSON_LoadEndInfo(
-            json_object_get_array(root_obj, "end_info"), game_info)) {
-        goto cleanup;
+    if (!Savegame_BSON_LoadResumeInfo(
+            json_object_get_array(root_obj, "current_info"),
+            game_info->current)) {
+        LOG_DEBUG("Failed to load current_info properly");
+        // Check for 2.6 and 2.7 legacy start and end info.
+        if (!Savegame_BSON_LoadDiscontinuedStartInfo(
+                json_object_get_array(root_obj, "start_info"), game_info)) {
```

Comment:
Now the issue is 2.6/2.7 saves cause a game crash.
```
../src/game/savegame_bson.c 173 Savegame_BSON_LoadResumeInfo Malformed save: invalid or missing resume array
../src/game/savegame_bson.c 1055 Savegame_BSON_LoadFromFile Failed to load current_info properly
```
Then black screen. Looks like this code does not properly revert to loading the legacy saves. I'm not as familiar with how the json loading works. `if (!Savegame_BSON_LoadResumeInfo(json_object_get_array(root_obj, "start_info"), game_info->start))` does not error. Does this throw off the file pointer maybe?

Reply:
JSON is all parsed at this point and you only traverse a virtual structure in the program's memory.
You need to ignore this error and provide fallback logic to have backwards compatibility with 2.6/2.7 saves.

Reply:
The idea was that if the `current_info` does not exist, try loading the legacy `start_info` and `end_info`. Will this logic not work?
```
    if (!Savegame_BSON_LoadResumeInfo(
            json_object_get_array(root_obj, "start_info"), game_info->start)) {
        LOG_DEBUG("Failed to load start_info properly");
        goto cleanup;
    }

    if (!Savegame_BSON_LoadResumeInfo(
            json_object_get_array(root_obj, "current_info"),
            game_info->current)) {
        LOG_DEBUG("Failed to load current_info properly");
        // Check for 2.6 and 2.7 legacy start and end info.
        if (!Savegame_BSON_LoadDiscontinuedStartInfo(
                json_object_get_array(root_obj, "start_info"), game_info)) {
            goto cleanup;
        }
        if (!Savegame_BSON_LoadDiscontinuedEndInfo(
                json_object_get_array(root_obj, "end_info"), game_info)) {
            goto cleanup;
        }
    }
```

Reply:
I think this should work.

Reply:
This is what I have and get the black screen. I'll have to debug the Discontinued functions more to see what's breaking.